File: src/server/services/risk.ts
import type { Application, Document, Requirement, Task } from "@prisma/client";
import type { RiskLevel } from "@/lib/enums";
import { daysUntil } from "@/lib/utils";
import { criticalIncomplete } from "./readiness";
import { daysWaiting, waitingRiskFromDays } from "./waiting";

type RiskInput = {
  applications: (Pick<
    Application,
    "hardDeadline" | "status" | "riskLevel"
  > & {
    requirements: Pick<Requirement, "status" | "isCritical">[];
  })[];
  documents: Pick<Document, "status" | "requestedAt">[];
  tasks: Pick<Task, "status" | "priority" | "dueDate">[];
};

const RANK: Record<RiskLevel, number> = {
  NONE: 0,
  LOW: 1,
  MEDIUM: 2,
  HIGH: 3,
  CRITICAL: 4,
};

function maxRisk(...levels: RiskLevel[]): RiskLevel {
  return levels.reduce((a, b) => (RANK[b] > RANK[a] ? b : a), "NONE" as RiskLevel);
}

export function calculateApplicationRisk(
  application: Pick<Application, "hardDeadline" | "status"> & {
    requirements: Pick<Requirement, "status" | "isCritical">[];
  },
  waitingDaysMax = 0,
  hasOverdueUrgent = false
): RiskLevel {
  if (
    ["SUBMITTED", "WAITING_RESULT", "ADMITTED", "REJECTED", "WAITLISTED", "ENROLLED", "NOT_SELECTED"].includes(
      application.status
    )
  ) {
    return "NONE";
  }

  const blockers = criticalIncomplete(application.requirements);
  const hasCriticalBlocker = blockers.length > 0;
  const incomplete = application.requirements.some(
    (r) => r.status !== "COMPLETED" && r.status !== "NOT_APPLICABLE"
  );
  const daysLeft = application.hardDeadline
    ? daysUntil(application.hardDeadline)
    : null;

  let level: RiskLevel = "NONE";

  if (daysLeft !== null && daysLeft <= 2 && hasCriticalBlocker) {
    level = maxRisk(level, "CRITICAL");
  }
  if (
    (daysLeft !== null && daysLeft <= 7 && hasCriticalBlocker) ||
    waitingDaysMax > 9 ||
    hasOverdueUrgent
  ) {
    level = maxRisk(level, "HIGH");
  }
  if (
    (daysLeft !== null && daysLeft <= 14 && incomplete) ||
    waitingDaysMax > 5
  ) {
    level = maxRisk(level, "MEDIUM");
  }
  if (incomplete) {
    level = maxRisk(level, "LOW");
  }

  return level;
}

export function calculateStudentRisk(input: RiskInput): RiskLevel {
  const waitingDocs = input.documents.filter(
    (d) => d.status === "REQUESTED" || d.status === "NEEDS_CHANGES"
  );
  const waitingDaysMax = waitingDocs.reduce(
    (max, d) => Math.max(max, daysWaiting(d.requestedAt)),
    0
  );
  const hasOverdueUrgent = input.tasks.some(
    (t) =>
      t.status !== "DONE" &&
      t.priority === "URGENT" &&
      t.dueDate &&
      daysUntil(t.dueDate) < 0
  );

  const appRisks = input.applications.map((a) =>
    calculateApplicationRisk(a, waitingDaysMax, hasOverdueUrgent)
  );

  const waitingRisk = waitingRiskFromDays(waitingDaysMax);
  return maxRisk(...appRisks, waitingRisk, hasOverdueUrgent ? "HIGH" : "NONE");
}

File: backend/app/services/operations/risk.py
from datetime import UTC, datetime

from app.services.operations.readiness import critical_incomplete

RISK_RANK = {"NONE": 0, "LOW": 1, "MEDIUM": 2, "HIGH": 3, "CRITICAL": 4}
CLOSED_APPLICATION_STATUSES = {
    "SUBMITTED",
    "WAITING_RESULT",
    "ADMITTED",
    "REJECTED",
    "WAITLISTED",
    "ENROLLED",
    "NOT_SELECTED",
}


def _days_until(deadline: datetime | None, now: datetime) -> int | None:
    if deadline is None:
        return None
    if deadline.tzinfo is None:
        deadline = deadline.replace(tzinfo=UTC)
    return (deadline.date() - now.date()).days


def calculate_application_risk(
    application_status: str,
    requirements: list[tuple[str, bool]],
    hard_deadline: datetime | None,
    waiting_days_max: int = 0,
    has_overdue_urgent: bool = False,
    now: datetime | None = None,
) -> str:
    if application_status in CLOSED_APPLICATION_STATUSES:
        return "NONE"
    now = now or datetime.now(UTC)
    days_left = _days_until(hard_deadline, now)
    has_blocker = critical_incomplete(requirements)
    incomplete = any(status not in {"COMPLETED", "NOT_APPLICABLE"} for status, _ in requirements)
    level = "NONE"

    def raise_to(candidate: str) -> None:
        nonlocal level
        if RISK_RANK[candidate] > RISK_RANK[level]:
            level = candidate

    if days_left is not None and days_left <= 2 and has_blocker:
        raise_to("CRITICAL")
    if (
        (days_left is not None and days_left <= 7 and has_blocker)
        or waiting_days_max >= 9
        or has_overdue_urgent
    ):
        raise_to("HIGH")
    if (days_left is not None and days_left <= 14 and incomplete) or waiting_days_max > 5:
        raise_to("MEDIUM")
    if incomplete:
        raise_to("LOW")
    return level


def calculate_student_risk_from_signals(
    application_risks: list[str],
    waiting_days_max: int,
    has_overdue_urgent: bool,
) -> str:
    waiting_risk = (
        "HIGH"
        if waiting_days_max >= 9
        else "MEDIUM"
        if waiting_days_max >= 6
        else "LOW"
        if waiting_days_max >= 3
        else "NONE"
    )
    levels = [*application_risks, waiting_risk]
    if has_overdue_urgent:
        levels.append("HIGH")
    return max(levels or ["NONE"], key=RISK_RANK.__getitem__)
